### PR TITLE
Feature/add agent version to installation directory name

### LINKED ISSUE #2579: Add the full version number to the installation directory name
## Dependencies ##
- Blocked by https://github.com/elastic/elastic-agent/issues/2873
- Blocked by https://github.com/elastic/elastic-agent/issues/2582
- Blocked by ttps://github.com/elastic/elastic-agent/issues/3639

## Issue ##

We need the full version number of agent to be present in the directory we install into. This is necessary for changing how we release agent.

## Suggested solution ##

Add the full version number to the installation directory name.
Example:
```
data/
└── elastic-agent-7ac68h-8.8.1
└── elastic-agent-7ac68h-8.8.1.er1
```

## Definition of done ##

- [ ] Installation directory now includes the full version number.
- [ ] Add a new integration test [here](https://github.com/elastic/elastic-agent/blob/main/testing/integration/install_test.go) to validate that installation directory is the correct one.

### PR DESCRIPTION
**Pull request by pchila**
## What does this PR do?
This PR uses the package manifest included in elastic-agent packages to map the package contents onto disk when installing/upgrading agent.
In the context of https://github.com/elastic/elastic-agent/issues/2579 the path mapping is used to install or upgrade agent in a directory like data/elastic-agent-<package version>-<hash> while we maintain backward compatibility for the package structure using the previous `elastic-agent-<hash>`.

This PR is the aggregation of #4173 #4174 #4176 #4177 #4178 that have been used to split the change in reviewable fragments.

With this change elastic-agent will use the path mappings present in the package manifest to remap package files before writing them on disk in the install directory.

There is no integration test in this change as that requires some hacking of the agent packages due to current limitations of integration test framework. This will be tackled in #2780 

## Why is it important?

This changes implements #2579 allowing different versions of agent built from the same git commit (for example a repackaged agent version containing patched components) to upgrade between themselves.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- ~~[ ] I have added an integration test or an E2E test~~

## Author's Checklist

- [ ]

## How to test this PR locally

To test this PR locally, we package elastic-agent twice from the same commit specifying a different AGENT_PACKAGE_VERSION like:
```
EXTERNAL=true SNAPSHOT=true PLATFORMS="linux/amd64" PACKAGES="tar.gz"  mage package
```
and
```
AGENT_PACKAGE_VERSION=8.13.0-repackaged EXTERNAL=true SNAPSHOT=true PLATFORMS="linux/amd64" PACKAGES="tar.gz"  mage package
```

Install the first agent package and verify that it's healthy and it runs from `data/elastic-agent-8.13.0-SNAPSHOT-<hash>` directory.
Upgrade this agent using the second package
```
elastic-agent upgrade 8.13.0-repackaged-SNAPSHOT --source-uri file://<directory where the package is located on disk> --skip-verify
```

Upgrade will unpack and store the new agent in `data/elastic-agent-8.13.0-repackaged-<hash>` and rotate the links.
Verify the agent directory structure contains the 2 distinct directories.

If we want to test the rollback functionality it's enough to kill the new agent a few times after the upgrade or break the new agent install in some way.

## Related issues

- Closes #2579
- Supersedes #3950 

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

**Review by ycombinator — APPROVED** [review]
Manually tested upgrading from `8.13.0-SNAPSHOT` -> `8.13.0-repackaged-SNAPSHOT` and roll back (by killing new Agent process a few times).  Checked folder structure on disk pre-upgrade / after successful upgrade / after roll back, and it looks as expected.  `elastic-agent version` and `elastic-agent status` commands report output as expected.  All looks good — nice job @pchila!

I did not do a code review as that was done for constituent PRs. 

One small thing I noticed, which could be improved/fixed in a follow up PR, is that during the upgrade from `8.13.0-SNAPSHOT` -> `8.13.0-repackaged-SNAPSHOT`, if I ran `sudo elastic-agent status`, the upgrade details showed the target version as just `8.13.0` (instead of `8.13.0-repackaged-SNAPSHOT`, which I think would be ideal).

```
$ sudo elastic-agent status
┌─ fleet
│  └─ status: (STOPPED) Not enrolled into Fleet
├─ elastic-agent
│  └─ status: (HEALTHY) Running
└─ upgrade_details
   ├─ target_version: 8.13.0
   ├─ state: UPG_WATCHING
   └─ metadata
```

**Comment by pchila** [comment]
> One small thing I noticed, which could be improved/fixed in a follow up PR, is that during the upgrade from `8.13.0-SNAPSHOT` -> `8.13.0-repackaged-SNAPSHOT`, if I ran `sudo elastic-agent status`, the upgrade details showed the target version as just `8.13.0` (instead of `8.13.0-repackaged-SNAPSHOT`, which I think would be ideal).

I noticed the same thing, I suspect that the upgrade details read the version not by calling `version.Version()` (source code below)

```go
// Version returns the version of the application.
func Version() string {
	return version.GetAgentPackageVersion()
}
```

**Comment by pchila** [comment]
buildkite test this

File: docs/upgrades.md [hunk 1]
```diff
@@ -61,3 +61,74 @@ sequenceDiagram
...
+#### Upgrading using the manifest
+
+Upgrading using the manifest allows for the new version to pass along some information about the package to the upgrading agent.
+The new process looks like this:
+- Download the elastic-agent package to use for upgrade
+- Extract package metadata from the new elastic-agent package (manifest and hash):
+  - if the package has a manifest we extract `version` and `snapshot` flag as declared by the package
+  - if there is no manifest for the package we extract the version and snapshot for the version passed to the upgrade
+  - hash from the agent commit file (always present in the package)
+- compare the tuple of new `(version, snapshot, hash)` to the current `(version, snapshot, hash)`: if they are the same
```
[the remark this answers is not in the document]
Comment (pchila) [reply docs/upgrades.md:125]: Yup that's a typo. The hash is always collected in the same way from the agent commit file, the difference between the manifest being in the package or not is in the way we extract `version` and `snapshot`. I will reword that

[the remark this answers is not in the document]
Comment (pchila) [reply docs/upgrades.md]: Have a look at 416c29d80df3fba7596a290a99d2bda3fa588dc0 and see if it's clearer

**Comment by cmacknz** [comment]
Even if we get a green build here, I think we should postpone merging this unless we can prove that https://github.com/elastic/elastic-agent/issues/4204 exists outside this branch.

**Comment by pchila** [comment]
@cmacknz 
I have disabled automerge.
 I would like to point out that the fact that some flakiness in our integration tests is brought out by this PR does not automatically mean that this PR has a bug (we would see more consistent failures of the same tests across multiple operating systems)
 I also would like to point out that this PR had green CI runs already before more commits were added to main, forcing a rebase of the branch

**Comment by pchila** [comment]
buildkite test this

**Comment by cmacknz** [comment]
> I would like to point out that the fact that some flakiness in our integration tests is brought out by this PR does not automatically mean that this PR has a bug (we would see more consistent failures of the same tests across multiple operating systems)
> I also would like to point out that this PR had green CI runs already before more commits were added to main, forcing a rebase of the branch

Yes CI is flaky without this change, but that we have a what appears to be a new failure mode that is isolated to this branch, whose consequence is unintentionally uninstalling the agent under some conditions, and we don't know why yet. To me this appears to be a new source of flakiness.

If we can determine this failure mode exists outside this branch, or we can understand what the failure mode specifically is, then we can move forward.

**Comment by cmacknz** [comment]
What I am specifically concerned about is the failure mode observed in https://github.com/elastic/elastic-agent/pull/4217#issuecomment-1933267945 where both the current and previous agent get removed by the watcher.

```json
{"log.level":"info","@timestamp":"2024-02-07T22:45:55.401Z","log.origin":{"file.name":"cmd/watch.go","file.line":67},"message":"Upgrade Watcher started","process.pid":24780,"agent.version":"8.13.0","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-02-07T22:45:55.406Z","log.origin":{"file.name":"upgrade/watcher.go","file.line":66},"message":"Agent watcher started","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-02-07T22:46:10.407Z","log.origin":{"file.name":"upgrade/watcher.go","file.line":133},"message":"Trying to connect to agent","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-02-07T22:46:10.408Z","log.origin":{"file.name":"upgrade/watcher.go","file.line":164},"message":"Connected to agent","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-02-07T22:46:10.411Z","log.origin":{"file.name":"upgrade/watcher.go","file.line":197},"message":"Communicating with PID 24739","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-02-07T22:46:54.901Z","log.origin":{"file.name":"cmd/watch.go","file.line":176},"message":"Grace period passed, not watching","ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2024-02-07T22:46:54.901Z","log.origin":{"file.name":"upgrade/watcher.go","file.line":184},"message":"Lost connection: failed reading next state: rpc error: code = Canceled desc = context canceled","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-02-07T22:46:54.906Z","log.origin":{"file.name":"upgrade/rollback.go","file.line":59},"message":"Cleaning up upgrade","hash":"8dcab2","remove_marker":true,"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-02-07T22:46:56.907Z","log.origin":{"file.name":"upgrade/step_mark.go","file.line":156},"message":"Removing marker file","file.path":"/opt/Elastic/Agent/data/.update-marker","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-02-07T22:46:56.907Z","log.origin":{"file.name":"upgrade/rollback.go","file.line":82},"message":"Removing previous symlink path","file.path":"/opt/Elastic/Agent/elastic-agent.prev","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-02-07T22:46:56.907Z","log.origin":{"file.name":"upgrade/rollback.go","file.line":97},"message":"Removing hashed data directory","file.path":"/opt/Elastic/Agent/data/elastic-agent-8.13.0-SNAPSHOT-b04c7c","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2024-02-07T22:46:57.198Z","log.origin":{"file.name":"upgrade/rollback.go","file.line":97},"message":"Removing hashed data directory","file.path":"/opt/Elastic/Agent/data/elastic-agent-8.13.0-SNAPSHOT-8dcab2","ecs.version":"1.6.0"}
```

That particular failure hasn't happened on main before AFAIK.

**Comment by cmacknz** [comment]
@pchila and I have discovered the source of the new test failure in the downgrade test:

- https://github.com/elastic/elastic-agent/issues/4228

We are working on a solution. The problem is only fatal when downgrading to a version before 8.13.0, and the way our tests work the failure will go away once we have a snapshot build that includes the changes in this PR. 

Depending on how long the fix for https://github.com/elastic/elastic-agent/issues/4228 takes we will have to discuss if we want to merge this PR without the fix, since it is a highly complex edge case that should not affect the officially supported upgrade flow.

**Comment by pchila** [comment]
buildkite test this

**Review by cmacknz — APPROVED** [review]
